ecosystem/incentivized-relays-mvp.md

```diff
@@ -0,0 +1,162 @@
+        uint256 amount = msg.value;
+        require(amount > 0);
+
+        uint256 newBalance = balanceOf[msg.sender] + amount;
+        require(newBalance < MAX_DEPOSIT);
+
+        balanceOf[msg.sender] = newBalance;
+    }
+}
+```
+
+We cap the deposits in order to cut scope in supporting withdrawals. This makes our first interation super simple since as withdrawal support introduces a race between deposited funds and a relayer compensating themselves for message delivery. With a relatively low max deposit, we eliminate the risk of a large amount of stuck funds for a given account whilst the amount being sufficient enough to cover hundreds of sub-cent transactions.
+
+Relayers that delivered messages can compensate themselves by pushing through the `RelayedMessageGasReceipt` event, validated with Superchain interop. A tx sender with no deposit makes this feature a no-op as no relayer will auto-deliver the message. The user but either relay messages themselves or can rely on a 3rdparty provider.
+
+```solidity
+contract L2ToL2CrossDomainGasTank {
+
+    mapping(bytes32 => bool) claimed;
+
+    function claim(Identifier calldata id, bytes calldata payload) nonReentrant external {
+        require(id.origin == address(messenger));
+        ICrossL2Inbox(Predeploys.CROSS_L2_INBOX).validateMessage(id, keccak256(payload));
+
+        // parse the receipt
+        require(payload[:32] == RelayedMessageGasReceipt.selector);
+        (bytes32 rootMsgHash, address txOrigin, uint256 callDepth, address relayer, uint256 relayCost) = _decode(payload);
+
+        // ensure message was sent from this chain
+        require(messenger.sentMessages[rootMsgHash]);
+
+        // ensure unclaimed, and mark the claim
+        bytes32 claimId = keccak256(abi.encode(rootMsgHash, callDepth));
+        require(!claimed[claimId]);
+
+        claimed[claimId] = true
+
+        // compute total cost
+        uint256 claimCost = CLAIM_OVERHEAD * block.basefee;
+        uint256 cost = relayCost + claimCost;
+        require(balanceOf[txOrigin] >= cost);
+
+        // compensate the relayer
+        balanceOf[txOrigin] -= cost;
+        new SafeSend{ value: cost }(payable(relayer));
+    }
+}
+```
+
+A claim is uniquely identified by the `(rootMsgHash, callDepth)` tuple.
+
+As relayers deliver messages, they can claim the cost against the original tx sender deposit with the gas tank. This design introduces some off-chain complexity for the relayer.
+
+1. The relayer must simulate the call and ensure the emitted tx origin has a deposit that will cover the cost on the originating chain.
+2. The relayer should also track the pending claimable `RelayedMessageGasReceipt` of the `rootMsgHash` callstack to maximize the likelihood that the held deposit is sufficient.
+3. The relayer should claim the receipt within a reasonable time frame to ensure they are compensated. An unrelated relayer not checking (2) might claim newer transitive message that depletes the funds.
```

Comment:
should we add `blockTimestamp` to `RelayedMessageGasReceipt` and let the gas tank owner set `maxReceiptAge`?